src/stcal/resample/resample.py

```diff
@@ -0,0 +1,1506 @@
+        nconpl = nmodels // 32 + (1 if nmodels % 32 else 0)
+        n_arr = self.n_arrays_per_output + 2  # 2 comes from pixmap
+        required_memory = npix * (n_arr * data_type.itemsize + nconpl * 4)
+
+        # compare used to available
+        used_fraction = required_memory / available_memory
+        if used_fraction > allowed_memory:
+            raise OutputTooLargeError(
+                f'Combined ImageModel size {self._output_wcs.array_shape} '
+                f'requires {bytes2human(required_memory)}. '
+                f'Model cannot be instantiated.'
+            )
+
+    def _compute_output_wcs(self, **wcs_pars):
+        """ returns a diustortion-free WCS object and its pixel scale """
+#        owcs = output_wcs_from_input_wcs(self._input_s_region_list, **wcs_pars)
+        owcs = output_wcs_from_input_wcs([self._input_img1_wcs], **wcs_pars)
+        return owcs['output_wcs'], owcs['pscale']
+
+    def preload_input_meta(self, wcs1, filename, s_region):
+        # set-up lists for WCS and file names
+        self._input_img1_wcs = None
+        self._input_s_region_list = []
+        self._input_filename_list = []
+
+        # loop over only science exposures in the ModelLibrary
+        # sci_indices = self._input_models.ind_asn_type("science")
+        with self._input_models:
+            for model in self._input_models:
+                # model = self._input_models.borrow(idx)
+
+                try:
+                    if self.get_model_attr_value(model, "exptype").upper() != "SCIENCE":
+                        self._input_models.shelve(model, modify=False)
+                        continue
+                except AttributeError:
+                    pass
+
+                if self._input_img1_wcs is None and wcs1:
+                    # extract all info needed from *this* model:
+                    self._input_img1_wcs = deepcopy(
+                        self.get_model_attr_value(model, "wcs")
+                    )
+                    self._input_img1_wcs.array_shape = self.get_model_array(
+                        model,
+                        "data"
+                    ).shape
+
+                if filename:
+                    self._input_filename_list.append(
+                        self.get_model_attr_value(model, "filename")
+                    )
+
+                if s_region:
+                    self._input_s_region_list.append(
+                        self.get_model_attr_value(model, "s_region")
+                    )
+
+                self._input_models.shelve(model, modify=False)
+
+        # store first model's entire meta (except for WCS and data):
+        if self._first_model_meta is None:
+            self._first_model_meta = self.new_model(copy_meta_from=model)
+
+    def build_driz_weight(self, model, weight_type=None, good_bits=None):
+        """Create a weight map for use by drizzle
+        """
+        data = self.get_model_array(model, "data")
```

Comment:
again here certain attributes of the model are assumed to exist, and IMO this should be made explicit either by having a model base class or by adding stcal dependencies (although we've been trying hard to avoid the latter)

Reply:
> again here certain attributes of the model are assumed to exist, and IMO this should be made explicit either by having a model base class or ... 

Seems like this is not going to happen. 

> ... by adding stcal dependencies (although we've been trying hard to avoid the latter)

I don't think this is necessary as this PR illustrates how to avoid imports.

Yes, it is true thst we assume some attributes to exist. I think for now we are safe but if you would like to harden the code, I could add `try-except` blocks around these data retrieval statements.